rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+            but instead of counts of tokens, will contain
+            the weighted count where weights are determined
+            by the ‘tfidf’ algorithm.
+            if, instead, mode is 'int', then the output vector is
+            an int tensor where each int is the index of one token
+            in the input string.
+        max_length:  None (default) | int.
+            Only used if mode=int. If set to an int,
+            the output int tensors are of shape [..., max_length],
+            with longer sequences being truncated at the end and
+            shorter sequences being right-padded.
+            If set to None, output sequences are
+            of shape [..., max_length_in_batch],
+            where max_length_in_batch is the length
+            of the longest sequence in the current batch:
+            shorter sequences get right-padded.
+
+    Input shape and type:
+        dtype: string.
+        shape: (batch_size, ..., 1)
+
+    Output shape and type:
+        if `mode='int'`:
+            dtype: int
+            shape: (batch_size, ..., max_length), where max_length
+                is the length of the longest token sequence in the current batch, or
+                the value of the argument `max_length` if it was passed.
+        else:
+            dtype: floating point.
+            shape: (batch_size, ..., num_tokens)
+
+    What happens in `adapt`:
+        We build an index mapping tokens to token indices,
+        and in the case of `mode='count'` and `mode='tfidf`,
+        we keep track of how many time each token has appeared.
+    """
+```
+
+### Writing a subclass of `PreprocessingLayer`
+
+The following 3 methods should be overridden:
+
+- `__init__`: constructor of the layer, used to configure its behavior.
+- `build(self, inputs_shape)`: creates the state variables of the layer.
+- `call(self, inputs)`: transforms the inputs (should only be called after `adapt` has been called).
+- `adapt(self, data, [reset_state=True])`: sets the state of the layer given the data provided (either as a tf.data dataset or numpy array(s)). The `reset_state` argument is optional and may be ignored.
+
+
+### Handling of async prefetching
+
+Some preprocessing ops are CPU-only and benefit from being executed asynchronously on the accelerator host (as opposed to the accelerator itself, e.g. GPU or TPU),
+with a batch of data being prepocessed on the host while the previous batch is being processed by the accelerator. This pattern is known as "async prefetching".
+
+This is normally done as part of a tf.data pipeline. The current proposal implies moving some of that preprocessing to inside the model itself, which is normally
+executed end-to-end on an accelerator.
+
+This means that we need a way to lift the preprocessing part of the model in a tf.data pipeline during model training. In `fit`, we can do this automatically.
+In custom training loops, we will expect the user to do it manually.
```

Comment:
Could you add an example?

Reply:
There is an example below ("Custom training loops" subsection).